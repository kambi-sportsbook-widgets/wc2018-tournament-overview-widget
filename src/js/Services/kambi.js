// import { getEventsByFilter, getEvent } from 'kambi-offering-api-module'
import { offeringModule } from 'kambi-widget-core-library'

/**
 * Fetches events by supplied filter
 * @param {String} filters Filters to check
 * @returns {Promise.<{filter: string, events: object[]}>}
 */
const getEvents = (filter, leftWidgetInput, rightWidgetInput) => {
  const dataRequests = [
    offeringModule.getEventsByFilter(filter), // tournament matches
    offeringModule.getEventsByFilter(`${filter}/all/all/competitions`), // tournament competitions (e.g. golden boot, tournament winner etc.)
  ]

  return Promise.all(dataRequests)
    .then(tournamentData => {
      if (tournamentData == null) {
        throw new Error(
          `No tournament data available for supplied filters: ${filter}, ${filter}/all/all/competitions`
        )
      }
      const tournamentEvents = tournamentData[0].events
      const tournamentCompetitions = tournamentData[1].events

      // get competitions to render by criterionId (hardcoded)
      let leftWidget
      let rightWidget
      tournamentCompetitions.forEach(competition => {
        competition.betOffers.forEach(betOffer => {
          if (competition.event.id === leftWidgetInput.eventId) {
            if (betOffer.criterion.id === leftWidgetInput.criterionId) {
              leftWidget = competition
            }
            return
          } else if (competition.event.id === rightWidgetInput.eventId) {
            if (betOffer.criterion.id === rightWidgetInput.criterionId) {
              rightWidget = competition
            }
            return
          }
          return
        })
      })

      if (!leftWidget || !rightWidget) {
        throw new Error(`Could not find matching criterion IDs from data`)
        return
      }

      // filter tournament events against main betoffer and correct criterion id as kambi sometimes adds other types of events to the same endpoint
      const tournamentMatches = tournamentEvents.filter(event => {
        let isTournamentMatch = false
        event.betOffers.forEach(offer => {
          if (offer.criterion.id === 1001159858 && offer.main) {
            isTournamentMatch = true
          }
        })
        return isTournamentMatch
      })

      return Promise.resolve({
        events: tournamentMatches,
        competitions: { leftWidget, rightWidget },
      })
    })
    .catch(err => {
      return Promise.reject(err)
    })
}

export const getMatchEvents = filter => {
  return offeringModule
    .getEventsByFilter(filter)
    .then(tournamentData => {
      if (tournamentData == null) {
        throw new Error(
          `No tournament data available for supplied filters: ${filter}, ${filter}/all/all/competitions`
        )
      }

      return Promise.resolve({ events: tournamentData.events })
    })
    .catch(err => {
      return Promise.reject(err)
    })
}

export default { getEvents }
