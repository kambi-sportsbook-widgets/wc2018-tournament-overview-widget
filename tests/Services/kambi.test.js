import kambi from '../../src/js/Services/kambi'
import {
  coreLibrary,
  offeringModule,
  widgetModule,
} from 'kambi-widget-core-library'

jest.mock('kambi-widget-core-library', () => ({
  coreLibrary: {
    config: {
      routeRoot: '',
    },
  },
  offeringModule: {
    getHighlight: jest.fn(),
    getEventsByFilter: jest.fn(),
  },
  widgetModule: {
    createFilterUrl: jest.fn(),
  },
}))

const mockLeftInput = { criterionId: 100, eventId: 999 }
const mockRightInput = { criterionId: 200, eventId: 666 }
const mockEvent = {
  event: {
    id: 999,
    group: 'VM 2018',
    groupId: 393929,
    name: 'Mock name',
    englishName: 'Mock english name',
    type: 'ET_MATCH',
  },
  liveData: {},
  betOffers: [
    { id: 1, outcomes: [], criterion: { id: mockLeftInput.criterionId } },
    { id: 2, outcomes: [], criterion: { id: mockRightInput.criterionId } },
  ],
}
const mockSecondEvent = {
  event: {
    id: 666,
    group: 'VM 2018',
    groupId: 393929,
    name: 'Mock name',
    englishName: 'Mock english name',
    type: 'ET_MATCH',
  },
  liveData: {},
  betOffers: [
    { id: 1, outcomes: [], criterion: { id: mockLeftInput.criterionId } },
    { id: 2, outcomes: [], criterion: { id: mockRightInput.criterionId } },
  ],
}

const mockCompetitions = {
  leftWidget: mockEvent,
  rightWidget: mockEvent,
}

describe('Kambi service progressive events', () => {
  beforeEach(() => {
    offeringModule.getEventsByFilter = jest.fn()
  })

  it('returns events correctly', () => {
    const filter = 'test/filter/1'

    offeringModule.getEventsByFilter.mockImplementation(() => {
      return new Promise(resolve =>
        resolve({ events: [mockEvent, mockSecondEvent] })
      )
    })

    return kambi
      .getEvents(filter, mockLeftInput, mockRightInput)
      .then(events => {
        expect(events).toMatchSnapshot()
        expect(offeringModule.getEventsByFilter).toHaveBeenCalledTimes(2)
      })
  })

  //   it('behaves correctly when no events were found', () => {
  //     const filter = 'test/filter/1'

  //     offeringModule.getEventsByFilter.mockReturnValueOnce(
  //       new Promise(resolve => resolve([{ events: [] }, { events: [] }]))
  //     )

  //     expect(offeringModule.getEventsByFilter).not.toHaveBeenCalled()

  //     return kambi
  //       .getEvents(filter)
  //       .then(events => {
  //         throw new Error('Not supposed to reach here')
  //       })
  //       .catch(err => {
  //         expect(
  //           err.message === `No events available for supplied filter: ${filter}`
  //         )
  //       })
  //   })
})
