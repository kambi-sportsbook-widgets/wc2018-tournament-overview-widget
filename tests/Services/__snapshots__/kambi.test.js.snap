// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Kambi service progressive events returns events correctly 1`] = `
Object {
  "competitions": Object {
    "leftWidget": Object {
      "betOffers": Array [
        Object {
          "criterion": Object {
            "id": 100,
          },
          "id": 1,
          "outcomes": Array [],
        },
        Object {
          "criterion": Object {
            "id": 200,
          },
          "id": 2,
          "outcomes": Array [],
        },
      ],
      "event": Object {
        "englishName": "Mock english name",
        "group": "VM 2018",
        "groupId": 393929,
        "id": 999,
        "name": "Mock name",
        "type": "ET_MATCH",
      },
      "liveData": Object {},
    },
    "rightWidget": Object {
      "betOffers": Array [
        Object {
          "criterion": Object {
            "id": 100,
          },
          "id": 1,
          "outcomes": Array [],
        },
        Object {
          "criterion": Object {
            "id": 200,
          },
          "id": 2,
          "outcomes": Array [],
        },
      ],
      "event": Object {
        "englishName": "Mock english name",
        "group": "VM 2018",
        "groupId": 393929,
        "id": 666,
        "name": "Mock name",
        "type": "ET_MATCH",
      },
      "liveData": Object {},
    },
  },
  "events": Array [],
}
`;
